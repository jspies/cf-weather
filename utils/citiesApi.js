const BASE_URL = 'http://localhost:3004/cities';

export const getCitiesMatchingQuery = async query => {
  const result = await fetch(`${BASE_URL}?name_like=${query}`);
  const cities = await result.json();
  return cities;
};
