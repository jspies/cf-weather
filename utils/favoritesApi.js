const BASE_URL = 'http://localhost:3004/favorites';

export const getFavorites = async () => {
  const result = await fetch(BASE_URL);
  const favorites = await result.json();
  return favorites;
};

export const addFavorite = async city => {
  const data = {
    city
  };
  await fetch(BASE_URL, {
    method: 'POST', // *GET, POST, PUT, DELETE, etc.
    mode: 'cors', // no-cors, cors, *same-origin
    cache: 'no-cache', // *default, no-cache, reload, force-cache, only-if-cached
    credentials: 'same-origin', // include, same-origin, *omit
    headers: {
      'Content-Type': 'application/json; charset=utf-8'
    },
    redirect: 'follow', // manual, *follow, error
    referrer: 'no-referrer', // no-referrer, *client
    body: JSON.stringify(data) // body data type must match "Content-Type" header
  });
};
