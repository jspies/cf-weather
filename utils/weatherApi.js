const APP_ID = 'd25483c53dd033e2c5de052978f9f9ee';
const BASE_URL = 'https://api.openweathermap.org/data/2.5';
const appendAppId = url => `${url}&appid=${APP_ID}`;

const makeApiCall = async (city, action) => {
  const url = appendAppId(`${BASE_URL}/${action}?q=${city}&units=Imperial`);
  const result = await fetch(url);
  const json = await result.json();
  return json;
};
export const getWeather = city => {
  return makeApiCall(city, 'weather');
};

export const getForecast = city => {
  return makeApiCall(city, 'forecast');
};
