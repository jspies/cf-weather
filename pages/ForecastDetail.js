import React from 'react';
import Link from 'next/link';
import ForecastLayout from '../components/ForecastLayout';
import { getForecast } from '../utils/weatherApi';
import {
  Table,
  TableHead,
  TableBody,
  TableRow,
  TableCell,
} from 'cf-component-table';
import { createComponent } from '@cloudflare/style-container';

const WeatherIcon = createComponent(
  () => ({
    width: '75px',
    height: '75px',
    borderRadius: '2px 2px 0 0',
    display: 'inline-block',
  }),
  'img',
  ['src']
);

const Title = createComponent(() => ({
  fontSize: '75px',
  textAlign: 'center',
}));

const Center = createComponent(() => ({
  fontSize: '30px',
  textAlign: 'center',
}));

const ForecastDetail = ({ cityName, weather }) => {
  return (
    <ForecastLayout>
      <Title>{cityName}</Title>
      <Table>
        <TableHead>
          <TableRow>
            <TableCell>
              Day
            </TableCell>
            <TableCell>
              Description
            </TableCell>
            <TableCell>
              High/Low
            </TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {weather.length > 0 ?
            weather.map((row, r) => (
              <TableRow key={r}>
                {row.map((cell, c) => (
                  <TableCell key={c}>
                    {cell}
                  </TableCell>
                ))}
              </TableRow>
            ))
            : null}
        </TableBody>
      </Table>
      <Center>
        <Link href={'/'}>
          <a>Back To Favorites</a>
        </Link>
      </Center>
    </ForecastLayout>
  );
};

/**
 * makes call for the forecast of the given city and sends as props to the stateless component
 * @param cityName - the city to get the forecast for
 */
ForecastDetail.getInitialProps = async ({ query: { cityName } }) => {
  const data = await getForecast(cityName);

  const weather = [];
  let currentDay;
  for (let i = 0; i < data.list.length; i++) {
    const day = data.list[i].dt_txt.split(' ')[0];
    const parts = day.split('-');
    if (currentDay !== day) {
      currentDay = day;
      weather.push([
        <div>
          <h3>{new Date(parts[0], parts[1] - 1, parts[2]).toDateString()}</h3>
          <WeatherIcon src={`http://openweathermap.org/img/w/${data.list[i].weather[0].icon}.png`} />
        </div>,
        data.list[i].weather[0].description,
        `${data.list[i].main.temp_max}°/${data.list[i].main.temp_min}°`,
      ]);
    }
  }

  return {
    cityName,
    weather,
  };
}

export default ForecastDetail;
