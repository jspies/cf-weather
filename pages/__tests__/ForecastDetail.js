jest.mock('../../utils/weatherApi', () => ({
  getForecast: jest.fn().mockImplementation(() => {
    return Promise.resolve({
      list: [
        {
          dt_txt: '2019-01-30',
          main: {
            temp_max: 23.14,
            temp_min: 21.87,
          },
          weather: [
            {
              icon: "01n",
              description: "clear sky",
            },
          ],
        },
        {
          dt_txt: '2019-01-31',
          main: {
            temp_max: 23.14,
            temp_min: 21.87,
          },
          weather: [
            {
              icon: "01n",
              description: "cloudy sky",
            },
          ],
        },
      ],
    });
  })
}));

import { felaShallow } from '@cloudflare/util-testing';
import ForecastDetail from '../ForecastDetail';
import { getForecast } from '../../utils/weatherApi';

describe('ForecastDetail', () => {
  let defaultProps;

  beforeEach(async () => {
    defaultProps = await ForecastDetail.getInitialProps({ query: { cityName: 'Austin' } });
  });

  it('should render the forecast details', done => {
    const { snapshot, wrapper } = felaShallow(
      <ForecastDetail {...defaultProps} />
    );

    setTimeout(() => {
      wrapper.update();
      expect(snapshot(wrapper)).toMatchSnapshot();
      done();
    });
  });
});
