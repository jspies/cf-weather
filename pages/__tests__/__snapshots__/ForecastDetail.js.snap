// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForecastDetail should render the forecast details 1`] = `
Object {
  "component": <ForecastLayout>
    <FelaComponent(div)>
      Austin
    </FelaComponent(div)>
    <Table
      bordered={true}
      empty={false}
      fixedLayout={true}
      loading={false}
      renderTableToDom={true}
      searching={false}
      translations={
        Object {
          "error": "Looks like something went wrong... Please try again later.",
          "loading": "Loading",
          "noItems": "There are currently no records.",
          "noItemsMatchingSearchTerm": "Hmm... We couldn't find any matching records",
        }
      }
    >
      <FelaComponent(TableHead)>
        <TableRow
          accent="none"
          hover={false}
        >
          <TableCell
            align="left"
            hideOnMobile={false}
            hover={false}
            textEllipsis={false}
          >
            Day
          </TableCell>
          <TableCell
            align="left"
            hideOnMobile={false}
            hover={false}
            textEllipsis={false}
          >
            Description
          </TableCell>
          <TableCell
            align="left"
            hideOnMobile={false}
            hover={false}
            textEllipsis={false}
          >
            High/Low
          </TableCell>
        </TableRow>
      </FelaComponent(TableHead)>
      <TableBody>
        <TableRow
          accent="none"
          hover={false}
          key="0"
        >
          <TableCell
            align="left"
            hideOnMobile={false}
            hover={false}
            key="0"
            textEllipsis={false}
          >
            <div>
              <h3>
                Wed Jan 30 2019
              </h3>
              <FelaComponent(img)
                src="http://openweathermap.org/img/w/01n.png"
              />
            </div>
          </TableCell>
          <TableCell
            align="left"
            hideOnMobile={false}
            hover={false}
            key="1"
            textEllipsis={false}
          >
            clear sky
          </TableCell>
          <TableCell
            align="left"
            hideOnMobile={false}
            hover={false}
            key="2"
            textEllipsis={false}
          >
            23.14째/21.87째
          </TableCell>
        </TableRow>
        <TableRow
          accent="none"
          hover={false}
          key="1"
        >
          <TableCell
            align="left"
            hideOnMobile={false}
            hover={false}
            key="0"
            textEllipsis={false}
          >
            <div>
              <h3>
                Thu Jan 31 2019
              </h3>
              <FelaComponent(img)
                src="http://openweathermap.org/img/w/01n.png"
              />
            </div>
          </TableCell>
          <TableCell
            align="left"
            hideOnMobile={false}
            hover={false}
            key="1"
            textEllipsis={false}
          >
            cloudy sky
          </TableCell>
          <TableCell
            align="left"
            hideOnMobile={false}
            hover={false}
            key="2"
            textEllipsis={false}
          >
            23.14째/21.87째
          </TableCell>
        </TableRow>
      </TableBody>
    </Table>
    <FelaComponent(div)>
      <Link
        href="/"
      >
        <a>
          Back To Favorites
        </a>
      </Link>
    </FelaComponent(div)>
  </ForecastLayout>,
  "styles": "

",
}
`;
